File: lib/turnbased/connect_four_ai.dart
import 'connect_four_model.dart';

/// Depth-limited minimax with alpha-beta pruning + simple line-score
/// heuristic at horizon. Connect Four state space is too big for full search;
/// depth 6 plays a reasonable medium-strength opponent.
int? bestMove(ConnectFourModel model, Disc ai, {int depth = 6}) {
  int bestScore = -1 << 30;
  int? bestCol;
  // Center-first ordering — strongest move historically.
  final order = _ordered(model.legalMoves().toList());
  for (final col in order) {
    final copy = model.copy();
    copy.apply(col);
    final s = _negamax(copy, ai.opponent, depth - 1, -1 << 30, 1 << 30, ai);
    if (s > bestScore) {
      bestScore = s;
      bestCol = col;
    }
  }
  return bestCol;
}

List<int> _ordered(List<int> moves) {
  moves.sort((a, b) =>
      (a - kCols ~/ 2).abs().compareTo((b - kCols ~/ 2).abs()));
  return moves;
}

int _negamax(ConnectFourModel m, Disc toMove, int depth, int alpha, int beta, Disc ai) {
  final w = m.winner;
  if (w != null) return w == ai ? 100000 - depth : -100000 + depth;
  if (m.isDraw) return 0;
  if (depth == 0) return _heuristic(m, ai);

  var best = -1 << 30;
  for (final col in _ordered(m.legalMoves().toList())) {
    final copy = m.copy();
    copy.apply(col);
    final s = -_negamax(copy, toMove.opponent, depth - 1, -beta, -alpha, ai);
    if (s > best) best = s;
    if (best > alpha) alpha = best;
    if (alpha >= beta) break;
  }
  return toMove == ai ? best : -best;
}

/// Score every length-4 window: +heavy if the AI has a near-complete line
/// with no opponent disc, -heavy if the opponent does.
int _heuristic(ConnectFourModel m, Disc ai) {
  int score = 0;
  for (var c = 0; c < kCols; c++) {
    for (var r = 0; r < kRows; r++) {
      for (final dir in const [
        [1, 0], [0, 1], [1, 1], [1, -1],
      ]) {
        if (!_inBounds(c + 3 * dir[0], r + 3 * dir[1])) continue;
        score += _scoreWindow(m, c, r, dir[0], dir[1], ai);
      }
    }
  }
  // Center column control matters.
  for (var r = 0; r < kRows; r++) {
    if (m.at(kCols ~/ 2, r) == ai) score += 3;
    if (m.at(kCols ~/ 2, r) == ai.opponent) score -= 3;
  }
  return score;
}

bool _inBounds(int c, int r) => c >= 0 && c < kCols && r >= 0 && r < kRows;

int _scoreWindow(ConnectFourModel m, int c, int r, int dc, int dr, Disc ai) {
  var mine = 0, theirs = 0, empty = 0;
  for (var i = 0; i < 4; i++) {
    final d = m.at(c + i * dc, r + i * dr);
    if (d == ai) {
      mine++;
    } else if (d == ai.opponent) {
      theirs++;
    } else {
      empty++;
    }
  }
  if (mine > 0 && theirs > 0) return 0;
  if (mine == 4) return 100;
  if (mine == 3 && empty == 1) return 10;
  if (mine == 2 && empty == 2) return 2;
  if (theirs == 3 && empty == 1) return -12;
  if (theirs == 2 && empty == 2) return -2;
  return 0;
}

File: ios/Ubapp/Games/ConnectFour/ConnectFourAI.swift
import Foundation

/// Depth-limited negamax with alpha-beta pruning + simple line-score
/// heuristic at horizon. Depth 6 plays a reasonable medium-strength opponent.
enum ConnectFourAI {
    static func bestMove(_ model: ConnectFourModel, ai: Disc, depth: Int = 6) -> Int? {
        var bestScore = Int.min
        var bestCol: Int?
        for col in ordered(model.legalMoves()) {
            var copy = model
            copy.apply(col)
            let s = negamax(copy, toMove: ai.opponent, depth: depth - 1,
                            alpha: Int.min / 2, beta: Int.max / 2, ai: ai)
            if s > bestScore { bestScore = s; bestCol = col }
        }
        return bestCol
    }

    private static func ordered(_ moves: [Int]) -> [Int] {
        moves.sorted { abs($0 - kCols / 2) < abs($1 - kCols / 2) }
    }

    private static func negamax(_ m: ConnectFourModel, toMove: Disc, depth: Int,
                                 alpha alphaIn: Int, beta: Int, ai: Disc) -> Int {
        if let w = m.winner { return w == ai ? 100_000 - depth : -100_000 + depth }
        if m.isDraw { return 0 }
        if depth == 0 { return heuristic(m, ai: ai) }
        var alpha = alphaIn
        var best = Int.min / 2
        for col in ordered(m.legalMoves()) {
            var copy = m
            copy.apply(col)
            let s = -negamax(copy, toMove: toMove.opponent, depth: depth - 1,
                             alpha: -beta, beta: -alpha, ai: ai)
            if s > best { best = s }
            if best > alpha { alpha = best }
            if alpha >= beta { break }
        }
        return toMove == ai ? best : -best
    }

    /// Score every length-4 window: heavy positive if the AI has a near-complete
    /// line with no opponent disc, heavy negative if the opponent does.
    private static func heuristic(_ m: ConnectFourModel, ai: Disc) -> Int {
        var score = 0
        let dirs = [(1,0),(0,1),(1,1),(1,-1)]
        for c in 0..<kCols {
            for r in 0..<kRows {
                for (dc, dr) in dirs {
                    if !inBounds(c + 3*dc, r + 3*dr) { continue }
                    score += scoreWindow(m, c: c, r: r, dc: dc, dr: dr, ai: ai)
                }
            }
        }
        // Center column control matters.
        for r in 0..<kRows {
            if m.at(kCols / 2, r) == ai { score += 3 }
            if m.at(kCols / 2, r) == ai.opponent { score -= 3 }
        }
        return score
    }

    private static func inBounds(_ c: Int, _ r: Int) -> Bool {
        c >= 0 && c < kCols && r >= 0 && r < kRows
    }

    private static func scoreWindow(_ m: ConnectFourModel, c: Int, r: Int,
                                     dc: Int, dr: Int, ai: Disc) -> Int {
        var mine = 0, theirs = 0, empty = 0
        for i in 0..<4 {
            let d = m.at(c + i * dc, r + i * dr)
            if d == ai { mine += 1 }
            else if d == ai.opponent { theirs += 1 }
            else { empty += 1 }
        }
        if mine > 0 && theirs > 0 { return 0 }
        if mine == 4 { return 100 }
        if mine == 3 && empty == 1 { return 10 }
        if mine == 2 && empty == 2 { return 2 }
        if theirs == 3 && empty == 1 { return -12 }
        if theirs == 2 && empty == 2 { return -2 }
        return 0
    }
}
